common/lib/xmodule/xmodule/seq_module.py

```diff
@@ -279,6 +282,79 @@ def verify_current_content_visibility(cls, date, hide_after_date):
             datetime.now(UTC) < date
         )
 
+    def gate_sequence_if_it_is_a_timed_exam_and_contains_content_type_gated_problems(self):
+        """
+        Problem:
+        Content type gating for FBE (Feature Based Enrollments) previously only gated individual blocks.
+        This was an issue because audit learners could start a timed exam
+        and then be unable to complete it because the graded content would be gated.
+        Even if they later upgraded, they could still be unable to complete the exam
+        because the timer could have expired.
+
+        Solution:
+        Gate the entire sequence when we think the above problem can occur.
+
+        If:
+        1. This sequence is a timed exam
+        2. And this sequence contains problems which this user cannot load due to content type gating
+        Then:
+        We will gate access to the entire sequence.
+        Otherwise, learners would have the ability to start their timer for an exam,
+        but then not have the ability to complete it.
+
+        We are displaying the gating fragment within the sequence, as is done for gating for prereqs,
+        rather than content type gating the entire sequence because that would remove the next/previous navigation.
+
+        When gated_sequence_fragment is not set to None, the sequence will be gated.
+
+        This functionality still needs to be replicated in the frontend-app-learning courseware MFE
+        The ticket to track this is https://openedx.atlassian.net/browse/REV-1220
+        Note that this will break compatability with using sequences outside of edx-platform
+        but we are ok with this for now
+        """
+        if not self.is_time_limited:
+            self.gated_sequence_fragment = None
+            return
+
+        try:
+            user = User.objects.get(id=self.runtime.user_id)
+            course_id = self.runtime.course_id
+            content_type_gating_service = self.runtime.service(self, 'content_type_gating')
+            if not (content_type_gating_service and
+                    content_type_gating_service.enabled_for_enrollment(user=user, course_key=course_id)):
+                self.gated_sequence_fragment = None
+                return
+
+            def leaf_filter(block):
+                # This function is used to check if this is a leaf block
+                # Blocks with children are not currently gated by content type gating
+                # Other than the outer function here
+                return (
+                    block.location.block_type not in ('chapter', 'sequential', 'vertical') and
```

Comment:
Why the explicit check for `block_type`, here? Would just using `not block.has_childen` work?

Reply:
i was copying from here
https://github.com/edx/edx-platform/blame/b13f207e1f5ec313998d08a106d7c0d2d4eee6c6/cms/djangoapps/contentstore/api/views/course_quality.py#L255
i feel like just block.has_children would work, though not sure if there are any edge cases to consider where you could have a vertical with no children for example?